In [97]:
import pandas as pd
import numpy as np
import matplotlib as plt

In [98]:
df = pd.read_csv("autodata.csv")

In [99]:
df.head(5)

Unnamed: 0  symboling  normalized-losses         make aspiration  \
0           0          3                122  alfa-romero        std   
1           1          3                122  alfa-romero        std   
2           2          1                122  alfa-romero        std   
3           3          2                164         audi        std   
4           4          2                164         audi        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   compression-ratio  horsepower  peak-rpm  city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0        21          27  13495.0   
1                9.0       111.0    5000.0        21          27  16500.0   
2                9.0       154.0    5000.0        19          26  16500.0   
3               10.0       102.0    5500.0        24          30  13950.0   
4                8.0       115.0    5500.0        18          22  17450.0   

   city-L/100km horsepower-binned  diesel  gas  
0     11.190476               Low       0    1  
1     11.190476               Low       0    1  
2     12.368421            Medium       0    1  
3      9.791667               Low       0    1  
4     13.055556               Low       0    1  

[5 rows x 30 columns]

In [100]:
df.tail(5)

Unnamed: 0  symboling  normalized-losses   make aspiration num-of-doors  \
196         196         -1                 95  volvo        std         four   
197         197         -1                 95  volvo      turbo         four   
198         198         -1                 95  volvo        std         four   
199         199         -1                 95  volvo      turbo         four   
200         200         -1                 95  volvo      turbo         four   

    body-style drive-wheels engine-location  wheel-base  ...  \
196      sedan          rwd           front       109.1  ...   
197      sedan          rwd           front       109.1  ...   
198      sedan          rwd           front       109.1  ...   
199      sedan          rwd           front       109.1  ...   
200      sedan          rwd           front       109.1  ...   

     compression-ratio  horsepower  peak-rpm  city-mpg highway-mpg    price  \
196                9.5       114.0    5400.0        23          28  16845.0   
197                8.7       160.0    5300.0        19          25  19045.0   
198                8.8       134.0    5500.0        18          23  21485.0   
199               23.0       106.0    4800.0        26          27  22470.0   
200                9.5       114.0    5400.0        19          25  22625.0   

     city-L/100km horsepower-binned  diesel  gas  
196     10.217391               Low       0    1  
197     12.368421            Medium       0    1  
198     13.055556            Medium       0    1  
199      9.038462               Low       1    0  
200     12.368421               Low       0    1  

[5 rows x 30 columns]

In [101]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 30 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         201 non-null    int64  
 1   symboling          201 non-null    int64  
 2   normalized-losses  201 non-null    int64  
 3   make               201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

In [102]:
df.describe()

Unnamed: 0   symboling  normalized-losses  wheel-base      length  \
count  201.000000  201.000000          201.00000  201.000000  201.000000   
mean   100.000000    0.840796          122.00000   98.797015    0.837102   
std     58.167861    1.254802           31.99625    6.066366    0.059213   
min      0.000000   -2.000000           65.00000   86.600000    0.678039   
25%     50.000000    0.000000          101.00000   94.500000    0.801538   
50%    100.000000    1.000000          122.00000   97.000000    0.832292   
75%    150.000000    2.000000          137.00000  102.400000    0.881788   
max    200.000000    3.000000          256.00000  120.900000    1.000000   

            width      height  curb-weight  engine-size        bore  \
count  201.000000  201.000000   201.000000   201.000000  201.000000   
mean     0.915126   53.766667  2555.666667   126.875622    3.330692   
std      0.029187    2.447822   517.296727    41.546834    0.268072   
min      0.837500   47.800000  1488.000000    61.000000    2.540000   
25%      0.890278   52.000000  2169.000000    98.000000    3.150000   
50%      0.909722   54.100000  2414.000000   120.000000    3.310000   
75%      0.925000   55.500000  2926.000000   141.000000    3.580000   
max      1.000000   59.800000  4066.000000   326.000000    3.940000   

           stroke  compression-ratio  horsepower     peak-rpm    city-mpg  \
count  197.000000         201.000000  199.000000   199.000000  201.000000   
mean     3.256904          10.164279  103.396985  5117.587940   25.179104   
std      0.319256           4.004965   37.553843   480.521824    6.423220   
min      2.070000           7.000000   48.000000  4150.000000   13.000000   
25%      3.110000           8.600000   70.000000  4800.000000   19.000000   
50%      3.290000           9.000000   95.000000  5200.000000   24.000000   
75%      3.410000           9.400000  116.000000  5500.000000   30.000000   
max      4.170000          23.000000  262.000000  6600.000000   49.000000   

       highway-mpg         price  city-L/100km      diesel         gas  
count   201.000000    201.000000    201.000000  201.000000  201.000000  
mean     30.686567  13207.129353      9.944145    0.099502    0.900498  
std       6.815150   7947.066342      2.534599    0.300083    0.300083  
min      16.000000   5118.000000      4.795918    0.000000    0.000000  
25%      25.000000   7775.000000      7.833333    0.000000    1.000000  
50%      30.000000  10295.000000      9.791667    0.000000    1.000000  
75%      34.000000  16500.000000     12.368421    0.000000    1.000000  
max      54.000000  45400.000000     18.076923    1.000000    1.000000

In [103]:
df.isnull()

Unnamed: 0  symboling  normalized-losses   make  aspiration  \
0         False      False              False  False       False   
1         False      False              False  False       False   
2         False      False              False  False       False   
3         False      False              False  False       False   
4         False      False              False  False       False   
..          ...        ...                ...    ...         ...   
196       False      False              False  False       False   
197       False      False              False  False       False   
198       False      False              False  False       False   
199       False      False              False  False       False   
200       False      False              False  False       False   

     num-of-doors  body-style  drive-wheels  engine-location  wheel-base  ...  \
0           False       False         False            False       False  ...   
1           False       False         False            False       False  ...   
2           False       False         False            False       False  ...   
3           False       False         False            False       False  ...   
4           False       False         False            False       False  ...   
..            ...         ...           ...              ...         ...  ...   
196         False       False         False            False       False  ...   
197         False       False         False            False       False  ...   
198         False       False         False            False       False  ...   
199         False       False         False            False       False  ...   
200         False       False         False            False       False  ...   

     compression-ratio  horsepower  peak-rpm  city-mpg  highway-mpg  price  \
0                False       False     False     False        False  False   
1                False       False     False     False        False  False   
2                False       False     False     False        False  False   
3                False       False     False     False        False  False   
4                False       False     False     False        False  False   
..                 ...         ...       ...       ...          ...    ...   
196              False       False     False     False        False  False   
197              False       False     False     False        False  False   
198              False       False     False     False        False  False   
199              False       False     False     False        False  False   
200              False       False     False     False        False  False   

     city-L/100km  horsepower-binned  diesel    gas  
0           False              False   False  False  
1           False              False   False  False  
2           False              False   False  False  
3           False              False   False  False  
4           False              False   False  False  
..            ...                ...     ...    ...  
196         False              False   False  False  
197         False              False   False  False  
198         False              False   False  False  
199         False              False   False  False  
200         False              False   False  False  

[201 rows x 30 columns]

In [104]:
df.isnull().sum()

Unnamed: 0           0
symboling            0
normalized-losses    0
make                 0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               4
compression-ratio    0
horsepower           2
peak-rpm             2
city-mpg             0
highway-mpg          0
price                0
city-L/100km         0
horsepower-binned    2
diesel               0
gas                  0
dtype: int64

In [105]:
df.notnull()

Unnamed: 0  symboling  normalized-losses  make  aspiration  num-of-doors  \
0          True       True               True  True        True          True   
1          True       True               True  True        True          True   
2          True       True               True  True        True          True   
3          True       True               True  True        True          True   
4          True       True               True  True        True          True   
..          ...        ...                ...   ...         ...           ...   
196        True       True               True  True        True          True   
197        True       True               True  True        True          True   
198        True       True               True  True        True          True   
199        True       True               True  True        True          True   
200        True       True               True  True        True          True   

     body-style  drive-wheels  engine-location  wheel-base  ...  \
0          True          True             True        True  ...   
1          True          True             True        True  ...   
2          True          True             True        True  ...   
3          True          True             True        True  ...   
4          True          True             True        True  ...   
..          ...           ...              ...         ...  ...   
196        True          True             True        True  ...   
197        True          True             True        True  ...   
198        True          True             True        True  ...   
199        True          True             True        True  ...   
200        True          True             True        True  ...   

     compression-ratio  horsepower  peak-rpm  city-mpg  highway-mpg  price  \
0                 True        True      True      True         True   True   
1                 True        True      True      True         True   True   
2                 True        True      True      True         True   True   
3                 True        True      True      True         True   True   
4                 True        True      True      True         True   True   
..                 ...         ...       ...       ...          ...    ...   
196               True        True      True      True         True   True   
197               True        True      True      True         True   True   
198               True        True      True      True         True   True   
199               True        True      True      True         True   True   
200               True        True      True      True         True   True   

     city-L/100km  horsepower-binned  diesel   gas  
0            True               True    True  True  
1            True               True    True  True  
2            True               True    True  True  
3            True               True    True  True  
4            True               True    True  True  
..            ...                ...     ...   ...  
196          True               True    True  True  
197          True               True    True  True  
198          True               True    True  True  
199          True               True    True  True  
200          True               True    True  True  

[201 rows x 30 columns]

In [106]:
df.notnull().sum()

Unnamed: 0           201
symboling            201
normalized-losses    201
make                 201
aspiration           201
num-of-doors         201
body-style           201
drive-wheels         201
engine-location      201
wheel-base           201
length               201
width                201
height               201
curb-weight          201
engine-type          201
num-of-cylinders     201
engine-size          201
fuel-system          201
bore                 201
stroke               197
compression-ratio    201
horsepower           199
peak-rpm             199
city-mpg             201
highway-mpg          201
price                201
city-L/100km         201
horsepower-binned    199
diesel               201
gas                  201
dtype: int64

Based on the summary above, each column has 201 rows of data, seven columns containing missing data:

stroke : 4 missing data

horsepower: 2 missing data

peak-rpm: 2 missing data

horsepower-binned: 2 missing data


In [107]:
# calculate mean value for stroke column 
avg_stroke = df["stroke"].astype("float").mean(axis = 0)
print("Average of stroke: ",avg_stroke)

# replace NaN(null) by mean value in "stroke" column
df["stroke"].replace(np.nan,avg_stroke, inplace = True)

Average of stroke:  3.256903553299492


Calculate the mean value for the 'horsepower' column:

In [108]:
avg_hp = df["horsepower"].astype("float").mean(axis = 0)
print("Average of horsepower:",avg_hp)
df["horsepower"].replace(np.nan,avg_hp,inplace = True)


Average of horsepower: 103.39698492462311


Calculate the mean value for the 'peakrpm' column:

In [109]:
avg_rpm = df["peak-rpm"].astype("float").mean(axis = 0)
print("Average of peak-rpm:", avg_rpm)

Average of peak-rpm: 5117.587939698493


In [110]:
df["peak-rpm"].replace(np.nan,avg_rpm,inplace= True)

In [111]:
df['num-of-doors'].value_counts()

num-of-doors
four    115
two      86
Name: count, dtype: int64

In [112]:
df['num-of-doors'].value_counts().idxmax()

'four'

In [113]:
#replace the missing 'num-of-doors' values by the most frequent 
df["num-of-doors"].replace(np.nan, "four", inplace=True)

# simply drop whole row with NaN in "horsepower-binned" column
df.dropna(subset=["horsepower-binned"], axis=0, inplace=True)

# reset index, because we dropped two rows
df.reset_index(drop=True, inplace=True)

In [114]:
df.isnull().sum()

Unnamed: 0           0
symboling            0
normalized-losses    0
make                 0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
city-L/100km         0
horsepower-binned    0
diesel               0
gas                  0
dtype: int64

In [115]:
df.head()

Unnamed: 0  symboling  normalized-losses         make aspiration  \
0           0          3                122  alfa-romero        std   
1           1          3                122  alfa-romero        std   
2           2          1                122  alfa-romero        std   
3           3          2                164         audi        std   
4           4          2                164         audi        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   compression-ratio  horsepower  peak-rpm  city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0        21          27  13495.0   
1                9.0       111.0    5000.0        21          27  16500.0   
2                9.0       154.0    5000.0        19          26  16500.0   
3               10.0       102.0    5500.0        24          30  13950.0   
4                8.0       115.0    5500.0        18          22  17450.0   

   city-L/100km horsepower-binned  diesel  gas  
0     11.190476               Low       0    1  
1     11.190476               Low       0    1  
2     12.368421            Medium       0    1  
3      9.791667               Low       0    1  
4     13.055556               Low       0    1  

[5 rows x 30 columns]

In [116]:
# Data standardization

df['highway-L/100km'] = 235/df["highway-mpg"]
df.head()

Unnamed: 0  symboling  normalized-losses         make aspiration  \
0           0          3                122  alfa-romero        std   
1           1          3                122  alfa-romero        std   
2           2          1                122  alfa-romero        std   
3           3          2                164         audi        std   
4           4          2                164         audi        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   horsepower  peak-rpm  city-mpg  highway-mpg    price city-L/100km  \
0       111.0    5000.0        21           27  13495.0    11.190476   
1       111.0    5000.0        21           27  16500.0    11.190476   
2       154.0    5000.0        19           26  16500.0    12.368421   
3       102.0    5500.0        24           30  13950.0     9.791667   
4       115.0    5500.0        18           22  17450.0    13.055556   

   horsepower-binned diesel  gas  highway-L/100km  
0                Low      0    1         8.703704  
1                Low      0    1         8.703704  
2             Medium      0    1         9.038462  
3                Low      0    1         7.833333  
4                Low      0    1        10.681818  

[5 rows x 31 columns]

In [117]:
# Data Normalization
# Target:would like to Normalize those variables so their value ranges from 0 to 1.
df['length'] = df['length']/df['length'].max()
df['width'] = df['width']/df['width'].max()

In [118]:
df['height'] = df['height']/df['height'].max() 
df[["length","width","height"]].head()

length     width    height
0  0.811148  0.890278  0.816054
1  0.811148  0.890278  0.816054
2  0.822681  0.909722  0.876254
3  0.848630  0.919444  0.908027
4  0.848630  0.922222  0.908027

In [119]:
# categorical to numerical variable conversion

In [120]:
df.columns

Index(['Unnamed: 0', 'symboling', 'normalized-losses', 'make', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price', 'city-L/100km', 'horsepower-binned', 'diesel',
       'gas', 'highway-L/100km'],
      dtype='object')

In [121]:
df['aspiration'].value_counts()

aspiration
std      163
turbo     36
Name: count, dtype: int64

In [122]:
dummy_variable_1 = pd.get_dummies(df["aspiration"],dtype = int)
dummy_variable_1.head()

std  turbo
0    1      0
1    1      0
2    1      0
3    1      0
4    1      0

In [123]:
df.head()

Unnamed: 0  symboling  normalized-losses         make aspiration  \
0           0          3                122  alfa-romero        std   
1           1          3                122  alfa-romero        std   
2           2          1                122  alfa-romero        std   
3           3          2                164         audi        std   
4           4          2                164         audi        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   horsepower  peak-rpm  city-mpg  highway-mpg    price city-L/100km  \
0       111.0    5000.0        21           27  13495.0    11.190476   
1       111.0    5000.0        21           27  16500.0    11.190476   
2       154.0    5000.0        19           26  16500.0    12.368421   
3       102.0    5500.0        24           30  13950.0     9.791667   
4       115.0    5500.0        18           22  17450.0    13.055556   

   horsepower-binned diesel  gas  highway-L/100km  
0                Low      0    1         8.703704  
1                Low      0    1         8.703704  
2             Medium      0    1         9.038462  
3                Low      0    1         7.833333  
4                Low      0    1        10.681818  

[5 rows x 31 columns]

In [124]:
df = pd.concat([df, dummy_variable_1], axis=1)
df.drop("aspiration", axis = 1, inplace=True)

In [125]:
df.head()

Unnamed: 0  symboling  normalized-losses         make num-of-doors  \
0           0          3                122  alfa-romero          two   
1           1          3                122  alfa-romero          two   
2           2          1                122  alfa-romero          two   
3           3          2                164         audi         four   
4           4          2                164         audi         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   

   city-mpg  highway-mpg    price city-L/100km horsepower-binned  diesel gas  \
0        21           27  13495.0    11.190476               Low       0   1   
1        21           27  16500.0    11.190476               Low       0   1   
2        19           26  16500.0    12.368421            Medium       0   1   
3        24           30  13950.0     9.791667               Low       0   1   
4        18           22  17450.0    13.055556               Low       0   1   

   highway-L/100km  std  turbo  
0         8.703704    1      0  
1         8.703704    1      0  
2         9.038462    1      0  
3         7.833333    1      0  
4        10.681818    1      0  

[5 rows x 32 columns]